src/main/java/org/squiddev/cobalt/lib/StringLib.java

```diff
@@ -654,6 +1048,192 @@ static Varargs sub(Varargs args) throws LuaError {
 		}
 	}
 
+	private static class UnpackIntState {
+		long result = 0;
+		int size = 0;
+	}
+
+	private static UnpackIntState unpackint(byte[] str, int offset, int size, ByteOrder endianness, int alignment, boolean signed) {
+		UnpackIntState retval = new UnpackIntState();
+		if (offset % Math.min(size, alignment) != 0 && alignment > 1) for (int i = 0; offset % Math.min(size, alignment) != 0 && i < alignment; i++) {offset++; retval.size++;}
+		for (int i = 0; i < size; i++) {
+			retval.result |= (((long) str[offset + i] & 0xFF) << ((endianness == ByteOrder.BIG_ENDIAN ? size - i - 1 : i) * 8));
+			retval.size++;
+		}
+		if (signed && (retval.result & (1 << (size * 8 - 1))) != 0)
+			for (int i = size; i < 8; i++) retval.result |= ((long)0xFF & 0xFF) << (i * 8);
+		return retval;
+	}
+
+	/**
+	 * string.unpack (fmt, s [, pos])
+	 *
+	 * Returns the values packed in string s (see string.pack) according to the format string fmt.
+	 * An optional pos marks where to start reading in s (default is 1).
+	 * After the read values, this function also returns the index of the first unread byte in s.
+	 */
+	static Varargs unpack(Varargs args) throws LuaError {
+		LuaString fmt = args.arg(1).checkLuaString();
```

Comment:
Likewise with the above, is it possible to chuck this into a `ByteBuffer` and use the various `.get*` methods to read it?

Reply:
The problem with using a `ByteBuffer` is that we need to be able to read arbitrary-length numbers, but `ByteBuffer` only provides methods to read set sizes (`char`, `int`, `long`, `float`, `double`). It would be much cleaner if we could use the preexisting conversion code that Java provides, but `string.pack` and `string.unpack` should be able to read 24-bit integers and the like if desired.

Reply:
Assumedly we can get away with this though, as there's still methods to read a single byte? I guess I'm just thinking it means we can avoid having to faff around with the buffer index.